File: frontend/schema/services/request.ts
import { z } from 'zod';
import { CURRENCIES } from '@/utils/global';

export const createRequestSchema = z
  .object({
    content: z
      .string()
      .min(4, 'validation_request_content_min_length')
      .max(512, 'validation_request_content_max_length'),
    budget: z
      .number()
      .min(0, 'validation_request_budget_min_length')
      .max(2147483647, 'validation_request_budget_max_length'),
    location: z
      .object({
        x: z.number().min(-180).max(180),
        y: z.number().min(-90).max(90)
      })
      .nullable(),
    radius: z.number().min(3000).max(1000000).nullable(),
    currency: z.enum(CURRENCIES as [string, ...string[]])
  })
  .refine(
    (data) =>
      (data.location !== null && data.radius !== null) ||
      (data.location === null && data.radius === null),
    {
      message: 'validation_request_location_radius'
    }
  );

export const editRequestSchema = z
  .object({
    content: z
      .string()
      .min(4, 'validation_request_content_min_length')
      .max(512, 'validation_request_content_max_length'),
    budget: z.number()
    .min(0, 'validation_request_budget_min_length')
    .max(2147483647, 'validation_request_budget_max_length'),
    location: z
      .object({
        x: z.number().min(-180).max(180),
        y: z.number().min(-90).max(90)
      })
      .nullable(),
    radius: z.number().min(3000).max(1000000).nullable()
  })
  .refine(
    (data) =>
      (data.location !== null && data.radius !== null) ||
      (data.location === null && data.radius === null),
    {
      message: 'validation_request_location_radius'
    }
  );

export const createAndEditOfferSchema = z.object({
  content: z
    .string()
    .min(4, 'validation_offer_content_min_length')
    .max(512, 'validation_offer_content_max_length'),
  price: z
    .number()
    .min(0, 'validation_offer_price_min_length')
    .max(2147483647, 'validation_offer_price_max_length'),
  negotiation: z.boolean().default(false)
});

export const requestByIdSchema = z.object({
  requestId: z
    .string()
    .refine(
      (value) => !isNaN(Number(value)),
      'requestId must be a valid number'
    )
    .transform((value) => Number(value))
});

File: mobile/lib/schemas/request.dart
import 'package:acanthis/acanthis.dart';

import '../utils/global.dart';

final createAndEditRequestSchema =
    object({
      'content': string()
          .min(4, message: "validation_request_content_min_length")
          .max(512, message: "validation_request_content_max_length"),
      'budget': number()
          .lte(2147483647, message: "validation_request_budget_max_length")
          .gte(0, message: "validation_request_budget_min_length"),
      'location': object({
        'x': number().gte(-180).lte(180),
        'y': number().gte(-90).lte(90),
      }).nullable(),
      'radius': number().gte(3000).lte(1000000).nullable(),
      'currency': string().enumerated(Currency.values),
    }).refine(
      onCheck: (value) {
        if ((value['location'] != null && value['radius'] != null) ||
            (value['location'] == null && value['radius'] == null)) {
          return true;
        }
        return false;
      },
      error: "validation_request_location_radius",
      name: 'message',
    );

final createAndEditOfferSchema = object({
  'content': string()
      .min(4, message: "validation_offer_content_min_length")
      .max(512, message: "validation_offer_content_max_length"),
  'price': number()
      .lte(2147483647, message: "validation_offer_price_max_length")
      .gte(0, message: "validation_offer_price_min_length"),
  'negotiation': boolean().nullable(defaultValue: false),
});
